src/provider.rs

```diff
@@ -313,6 +375,81 @@ impl Future for Provider {
     }
 }
 
+struct Handler(Database);
+
+impl Handler {
+    fn list(self, _msg: ListRequest) -> impl Stream<Item = ListResponse> + Send + 'static {
+        let items = self
+            .0
+            .blobs()
+            .into_iter()
+            .map(|(hash, path, size)| ListResponse { hash, path, size });
+        futures::stream::iter(items)
+    }
+    async fn provide(self, msg: ProvideRequest) -> anyhow::Result<ProvideResponse> {
```

Comment:
Maybe call this the same as on the cli where it's called `add` (and the request type).  I think that's more cohesive.